mlflow/pyfunc/__init__.py

```diff
@@ -828,7 +828,46 @@ def _create_model_downloading_tmp_dir(should_use_nfs):
 _MLFLOW_SERVER_OUTPUT_TAIL_LINES_TO_KEEP = 200
 
 
-def spark_udf(spark, model_uri, result_type="double", env_manager=_EnvManager.LOCAL):
+def _infer_spark_udf_return_type(model_output_schema):
+    from pyspark.sql.types import StructType, StructField, ArrayType
+
+    def _convert_spec_to_spark_type(spec):
```

Comment:
Does this function have to be an inner function? IMO, a function like this evil and makes the code difficult to read.